core/primitives/src/math.rs

```diff
@@ -0,0 +1,392 @@
+    Ratio::new(result, d as i128)
+}
+
+/// Computes exp(x) - 1
+fn expm1(x: Ratio<i128>) -> Ratio<i128> {
+    // Short circuit extreme inputs.
+    let whole_part = x.numer() / x.denom();
+    if whole_part < -29 {
+        return Ratio::from_integer(-1);
+    } else if whole_part > 88 {
+        return Ratio::from_integer(i128::MAX);
+    }
+
+    // Keep size of denominator ~= PRECISION
+    let (n, d) = round_to_precision(x, PRECISION);
+
+    let result = iexpm1(n, d as u128);
+    Ratio::new(result, d)
+}
+
+trait SaturatingMul {
+    fn saturating_mul(self, other: Self) -> Self;
+}
+
+impl SaturatingMul for u128 {
+    fn saturating_mul(self, other: Self) -> Self {
+        u128::saturating_mul(self, other)
+    }
+}
+
+impl SaturatingMul for i128 {
+    fn saturating_mul(self, other: Self) -> Self {
+        i128::saturating_mul(self, other)
+    }
+}
+
+/// Used to keep the denominator of the fractions passed to
+/// `exp` and `ln` of order `precision`. This prevents accuracy
+/// problems introduced by overflow.
+fn round_to_precision<T: Copy + Num + Ord + FromPrimitive + SaturatingMul>(
+    x: Ratio<T>,
+    precision: T,
+) -> (T, T) {
+    let n = *x.numer();
+    let d = *x.denom();
+    let ten = T::from_u32(10).unwrap();
+
+    if d < precision {
+        let f = precision / d;
+        (f.saturating_mul(n), f.saturating_mul(d))
+    } else if d > ten * precision {
+        let f = d / precision;
+        (n / f, d / f)
+    } else {
+        (n, d)
+    }
+}
+
+/// When using Taylor Series to compute iexpm1 or ilnp1, require `|input| <= 2^{-TAYLOR_BOUND}`.
+const TAYLOR_BOUND: u32 = 8;
+
+/// Computes `denom * expm1(numer / denom)` to the nearest integer, where `expm1(x) = exp(x) - 1`.
+/// Overflows are handled using saturating operations, this means the result may be unreliable if
+/// attempting to obtain too much precision (i.e. `numer` and `denom` are both very large).
+fn iexpm1(numer: i128, denom: u128) -> i128 {
+    // If the denominator overflows i128, then drop the least significant bit from both
+    // numerator and denominator.
+    let (numer, denom) = if denom > (i128::MAX as u128) {
+        (numer >> 1, (denom >> 1) as i128)
+    } else {
+        (numer, denom as i128)
+    };
+
+    let abs_numer = abs(numer);
+
+    // Find a value of `r` such that `|numer / denom| * 2^{-r} < 2^{-TAYLOR_BOUND}`
+    let r = (TAYLOR_BOUND + next_power_of_two(abs_numer))
+        .saturating_sub(next_power_of_two(denom as u128));
+
+    // Use the Taylor series to compute 2^{r} * denom * expm1(numer / (2^r * denom))
+    let scaled_denom = saturating_shl(denom, r);
+    let mut taylor_result = 0i128;
+    let mut curr_term = numer;
+    let mut term_count = 1;
+    while curr_term.abs() > 0 {
+        taylor_result += curr_term;
+        term_count += 1;
+        curr_term =
+            div_nearest(curr_term.saturating_mul(numer), scaled_denom.saturating_mul(term_count));
+        if term_count > 100 {
```

Comment:
How did you choose 100?

Reply:
The value doesn't really matter. Its purpose is to just bound the loop in some denigrate cases (only when near overflow). In the vast majority of cases the loop stops because of the `while` condition after like 6 terms (the Taylor series coverages very quickly for small inputs, which is why we use the identities mentioned in the comments).